### PR TITLE
ddl: invalid multiple MAXVALUE partitions (#36329)

### LINKED ISSUE #36329: PARTITION BY RANGE COLUMNS allows multiple MAXVALUE partitions
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t (d datetime) partition by range columns (d) (partition p1 values less than ('2022-01-01'), partition p2 values less than (maxvalue), partition p3 values less than (MAXVALUE));
```

### 2. What did you expect to see? (Required)
```
mysql>  create table t (d datetime) partition by range columns (d) (partition p1 values less than ('2022-01-01'), partition p2 values less than (maxvalue), partition p3 values less than (MAXVALUE));
ERROR 1493 (HY000): VALUES LESS THAN value must be strictly increasing for each partition
```

### 3. What did you see instead (Required)
```
tidb> create table t (d datetime) partition by range columns (d) (partition p1 values less than ('2022-01-01'), partition p2 values less than (maxvalue), partition p3 values less than (MAXVALUE));
Query OK, 0 rows affected (0,00 sec)

tidb> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `d` datetime DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE COLUMNS(`d`)
(PARTITION `p1` VALUES LESS THAN ('2022-01-01'),
 PARTITION `p2` VALUES LESS THAN (MAXVALUE),
 PARTITION `p3` VALUES LESS THAN (MAXVALUE))
1 row in set (0,00 sec)
```

### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v6.2.0-alpha-378-g987bdd3014
Edition: Community
Git Commit Hash: 987bdd3014d8c90302caac3e99092d8014c47109
Git Branch: master
UTC Build Time: 2022-07-19 09:33:05
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
Store: unistore
```

### PR DESCRIPTION
**Pull request by u5surf**
Signed-off-by: u5surf <u5.horie@gmail.com>

### What problem does this PR solve?

Issue Number: close #36329 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

Release-note
```release-note
None
```

### CODE REVIEW

**Review by mjonss — APPROVED** [review]
LGTM for single column RANGE COLUMNS (which is what we currently support), but it will not work for multi-column RANGE COLUMNS, like this:
```sql
create table t (a int, b int, c int) partition by range columns (a,b,c)
(partition p0 values less than (10, MAXVALUE, 0),
 partition p1 values less than (10, MAXVALUE, 1000))
```
(which also fails in MySQL, but I consider it to be a bug.)

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -2762,6 +2762,10 @@ func checkTwoRangeColumns(ctx sessionctx.Context, curr, prev *model.PartitionDef
 	for i := 0; i < len(pi.Columns); i++ {
 		// Special handling for MAXVALUE.
 		if strings.EqualFold(curr.LessThan[i], partitionMaxValue) {
+			if strings.EqualFold(prev.LessThan[i], partitionMaxValue) {
+				// Current and previous are both maxvalue.
+				return false, nil
+			}
 			// If current is maxvalue, it certainly >= previous.
 			return true, nil
```
[the remark this answers is not in the document]
Comment (u5surf) [reply ddl/ddl_api.go]: @mjonss fixed

File: ddl/db_partition_test.go [hunk 2]
```diff
@@ -563,6 +563,14 @@ create table log_message_1 (
 				"partition p1 values less than ('A'));",
 			dbterror.ErrRangeNotIncreasing,
 		},
+		{
+			"create table t(d datetime)" +
+				"partition by range columns (d) (" +
+				"partition p0 values less than ('2022-01-01')," +
+				"partition p1 values less than (MAXVALUE), " +
+			        "partition p2 values less than (MAXVALUE));",
```
Comment (mjonss) [inline ddl/db_partition_test.go:571]: Please format this line, I think you can just do `make dev` and it will reformat it for you.

**Review by mjonss — APPROVED** [review]

File: ddl/db_partition_test.go (hunk 2 above)
Comment (u5surf) [reply ddl/db_partition_test.go]: @mjonss reformatted with make dev

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -2762,8 +2762,10 @@ func checkTwoRangeColumns(ctx sessionctx.Context, curr, prev *model.PartitionDef
 	for i := 0; i < len(pi.Columns); i++ {
 		// Special handling for MAXVALUE.
 		if strings.EqualFold(curr.LessThan[i], partitionMaxValue) {
```
Comment (Defined2014) [inline ddl/ddl_api.go:2764]:
```suggestion
		if strings.EqualFold(curr.LessThan[i], partitionMaxValue) && !strings.EqualFold(prev.LessThan[i], partitionMaxValue) {
```

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -2762,8 +2762,10 @@ func checkTwoRangeColumns(ctx sessionctx.Context, curr, prev *model.PartitionDef
...
-			// If current is maxvalue, it certainly >= previous.
-			return true, nil
+			if !strings.EqualFold(prev.LessThan[i], partitionMaxValue) {
```
Comment (Defined2014) [inline ddl/ddl_api.go:2765]: I think combine 2764 and 2765 in one line is better.

Comment (u5surf) [reply ddl/ddl_api.go]: @Defined2014 Exactly, fixed.

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -2761,9 +2761,10 @@ func checkTwoRangeColumns(ctx sessionctx.Context, curr, prev *model.PartitionDef
 	}
 	for i := 0; i < len(pi.Columns); i++ {
 		// Special handling for MAXVALUE.
-		if strings.EqualFold(curr.LessThan[i], partitionMaxValue) {
+		if strings.EqualFold(curr.LessThan[i], partitionMaxValue) && !strings.EqualFold(prev.LessThan[i], partitionMaxValue) {
 			// If current is maxvalue, it certainly >= previous.
 			return true, nil
+
```
[the remark this answers is not in the document]
Comment (u5surf) [reply ddl/ddl_api.go]: @Defined2014 my bad.. Fixed.

**Comment by u5surf** [comment]
Hi, @Defined2014. The CI failed alert won't stop ringing now, what should I do?

**Comment by Defined2014** [comment]
@u5surf I think you can ignore this. We cherry-pick this commit to other release branch with broken CI.

**Pull request merged**